src/preferences/configobject.cpp

```diff
@@ -186,34 +187,65 @@ template <class ValueType> void ConfigObject<ValueType>::reopen(const QString& f
     }
 }
 
-template <class ValueType> void ConfigObject<ValueType>::save() {
+/// Save the ConfigObject to disk.
+/// Returns true on success
+template<class ValueType>
+bool ConfigObject<ValueType>::save() {
     QReadLocker lock(&m_valuesLock); // we only read the m_values here.
-    QFile file(m_filename);
-    if (!QDir(QFileInfo(file).absolutePath()).exists()) {
-        QDir().mkpath(QFileInfo(file).absolutePath());
+    QFile tmpFile(m_filename + kTempFilenameExtension);
+    if (!QDir(QFileInfo(tmpFile).absolutePath()).exists()) {
+        QDir().mkpath(QFileInfo(tmpFile).absolutePath());
+    }
+    if (!tmpFile.open(QIODevice::WriteOnly | QIODevice::Text)) {
+        qWarning() << "Could not write config file: " << tmpFile.fileName();
+        return false;
+    }
+    QTextStream stream(&tmpFile);
+    stream.setCodec("UTF-8");
+
+    QString group = "";
+
+    qint64 minLength = 0;
+    for (auto i = m_values.constBegin(); i != m_values.constEnd(); ++i) {
+        //qDebug() << "group:" << it.key().group << "item" << it.key().item << "val" << it.value()->value;
+        if (i.key().group != group) {
+            group = i.key().group;
+            stream << "\n"
+                   << group << "\n";
+            minLength += i.key().group.length() + 2;
+        }
+        stream << i.key().item << " " << i.value().value << "\n";
+        minLength += i.key().item.length() + i.value().value.length() + 1;
     }
-    if (!file.open(QIODevice::WriteOnly | QIODevice::Text)) {
-        qDebug() << "Could not write file" << m_filename << ", don't worry.";
-        return;
-    } else {
-        QTextStream stream(&file);
-        stream.setCodec("UTF-8");
 
-        QString grp = "";
+    stream.flush();
+    // the stream is usually longer, depending on the amount of encoded data.
+    if (stream.pos() < minLength || QFileInfo(tmpFile).size() != stream.pos()) {
```

Comment:
Do we really need the redundant minLength check if the stream already keeps track of the position? Why and how could the stream position ever fall behind minLength??

It adds a lot of redundant code that needs to be maintained and it is unknown if it ever catches any inconsistency.

Reply:
Either take it or leave it now. I will not put any more work in this now. It works, it is safe for all cases I could test. Even partial writes of the file.

Reply:
Without the minLength check you need to check some other way for the safeConfig object to not cause errors being reported when writing the empty config file.

Reply:
I still don't get the reasoning. Why is checking QTextStream::status() for QTextStream::WriteFailed after flushing not sufficient to detect any write errors?

Reply:
Because there is a safemode instance running that writes a 0 byte file by default and triggers the error handler. Instead of setting a bool from 0 to 1 I decided to do extra safety and count what is at least expected. For me, loosing settings is super annoying, so I usually put them to a safe space. I have decades of experience with computer systems to know that a full file systems behaviour depends on the system implementation.
This is the nit picking be mentioned before, that is also bothering me very much.

Reply:
@daschuer @Be @Holzhaus Could somone else please comment on this? Does the code reveal all those thoughts? I suppose that any other developer stumbling over this code in the future will ask the same questions I was asking.

We rely on Qt for doing the system abstractions for us. If this doesn't work as expected and we need to introduce workarounds that are basically mistrusting and bypassing the public Qt API, then we should document those exceptional behaviors and the corresponding mitigations clearly.

Reply:
I agree, there should at least be a comment why this is necessary in the code. Otherwise the next dev that touches this might just remove it, because it looks redundant.

Reply:
Thanks for adding a comment!

Recommended reading (off the record): http://danluu.com/file-consistency/

TL;DR: Use SQLite as a good default if you need to store data reliably.